src/Hazelcast.Net/Serialization/Compact/CompactDictionaryGenericRecord.cs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+// http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Hazelcast.Serialization.Compact;
+
+/// <summary>
+/// Implements <see cref="IGenericRecord"/> for compact serialization with
+/// values being kept in a dictionary.
+/// </summary>
+internal partial class CompactDictionaryGenericRecord : CompactGenericRecordBase, IGenericRecord
+{
+    private readonly IDictionary<string, object?> _fieldValues;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="CompactDictionaryGenericRecord"/> class.
+    /// </summary>
+    /// <param name="schema">The compact schema for the record.</param>
+    /// <param name="fieldValues">The values for the fields.</param>
+    public CompactDictionaryGenericRecord(Schema schema, IDictionary<string, object?> fieldValues)
+        : base(schema)
+    {
+        _fieldValues = fieldValues ?? throw new ArgumentNullException(nameof(fieldValues));
+    }
+
+    /// <inheritdoc />
+    public IGenericRecordBuilder NewBuilder() => new CompactGenericRecordBuilder(Schema);
+
+    /// <inheritdoc />
+    public IGenericRecordBuilder NewBuilderWithClone() => new CompactGenericRecordBuilder(Schema, _fieldValues);
+
+    private void ValidateField(string fieldname, FieldKind kind, FieldKind altKind = FieldKind.NotAvailable)
+    {
+        if (!Schema.TryGetField(fieldname, out var field) || (field.Kind != kind && field.Kind != altKind)) 
+            throw new SerializationException($"Record does not contain a field named '{fieldname}' of the requested {kind} kind.");
+    }
+
+    /// <inheritdoc />
+    public override object? GetFieldValue(string fieldname) => _fieldValues[fieldname];
+
+    /// <inheritdoc />
+    public IGenericRecord? GetGenericRecord(string fieldname)
+    {
+        ValidateField(fieldname, FieldKind.Compact);
+        return (IGenericRecord?) _fieldValues[fieldname];
+    }
+
+    /// <inheritdoc />
+    public IGenericRecord?[]? GetArrayOfGenericRecord(string fieldname)
+    {
+        ValidateField(fieldname, FieldKind.ArrayOfCompact);
+        return (IGenericRecord?[]?)_fieldValues[fieldname];
+    }
+
+    private static T[]? GetArrayOf<T>(object? obj) 
+        where T : struct
+    {
+        if (obj == null) return null;
```

Comment:
Don't know .net magic tricks but is this line necessary? If not, you can remove it from the other method as well.

Reply:
let's say it's magic - we need it